File: 0973-k-closest-points-to-origin/0973-k-closest-points-to-origin.java
class Solution {
    public int distanceFromOrigin(int[] point) {
        return point[1] * point[1] + point[0] * point[0];
    }
    public int[][] kClosest(int[][] points, int k) {
        PriorityQueue<int[]> maxHeap = new PriorityQueue<int[]>((int[] a, int[] b) -> {
            return distanceFromOrigin(b) - distanceFromOrigin(a);
        });
       
        for(int[] point : points) {
            if(maxHeap.size() == k) {
                if(distanceFromOrigin(maxHeap.peek()) > distanceFromOrigin(point)) {
                    maxHeap.poll();
                    maxHeap.offer(point);
                }
            } else {
                maxHeap.offer(point);
            }
        }
        int[][] res = new int[k][2];
        while(--k >= 0) {
            res[k] = maxHeap.poll();
        }
        
        return res;
    }
}

File: 0973-k-closest-points-to-origin/0973-k-closest-points-to-origin.cpp
class Solution {
public:
    vector<vector<int>> kClosest(vector<vector<int>>& points, int k) {
        auto cmp = [](vector<int> a, vector<int> b) {
            return a[0]*a[0] + a[1]*a[1] < b[0]*b[0] + b[1]*b[1];
        };
        priority_queue<vector<int>, vector<vector<int>>, decltype(cmp)> pq(cmp);
        vector<vector<int>> res(k);
        for(auto point: points) {
            pq.push(point);
            if(pq.size() > k)
                pq.pop();
        }
        while(!pq.empty()) {
            res[--k] = pq.top();
            pq.pop();
        }
        return res;
    }
};